File: week5/JH429/BOJ_1991.c
#include <stdio.h>

typedef struct{
    char parent, leftChild, rightChild;
} Node;

Node nodes[27];

void preorder(char root){
    if (root == '.')
        return;
    
    printf("%c", root);
    preorder(nodes[root - 'A'].leftChild);
    preorder(nodes[root - 'A'].rightChild);
}

void inorder(char root){
    if (root == '.')
        return;
    
    inorder(nodes[root - 'A'].leftChild);
    printf("%c", root);
    inorder(nodes[root - 'A'].rightChild);
}

void postorder(char root){
    if (root == '.')
        return;
    
    postorder(nodes[root - 'A'].leftChild);
    postorder(nodes[root - 'A'].rightChild);
    printf("%c", root);
}


int main() {
    int N;
    char p, l, r;

    scanf("%d", &N);
    getchar();

    for(int i = 0; i < N; i++) {
        scanf("%c %c %c", &p, &l, &r);
        getchar();
        nodes[p - 'A'].parent = p;
        nodes[p - 'A'].leftChild = l;
        nodes[p - 'A'].rightChild = r;
    }

    preorder('A');
    printf("\n");
    inorder('A');
    printf("\n");
    postorder('A');

    return 0;
    
}

File: week5/GamGomYang/BOJ1991.py
#1. 입력 노드 수와 노드 정보
n = int(input())
tree = {}
for _ in range(n):
    node , left, right = list(input().split())
    tree[node] = left, right


#2. 전위 순회
def preorder(x):
    if x == '.':
        return
    print(x, end='')      # 현재 노드 방문
    preorder(tree[x][0])  # 왼쪽 자식 방문
    preorder(tree[x][1])  # 오른쪽 자식 방문


#3. 중위 순회
def inorder(x):
    if x == '.':
        return
    inorder(tree[x][0])  # 왼쪽 자식 방문
    print(x, end='')     # 현재 노드 방문
    inorder(tree[x][1])  # 오른쪽 자식 방문


#4. 후위 순회
def postorder(x):
    if x == '.':
        return
    postorder(tree[x][0])  # 왼쪽 자식 방문
    postorder(tree[x][1])  # 오른쪽 자식 방문
    print(x, end='')       # 현재 노드 방문


preorder('A')
print()
inorder('A')
print()
postorder('A')